src/msg/protocol/proto/decoder.go

```diff
@@ -24,27 +24,35 @@ import (
 	"fmt"
 	"io"
 
+	xio "github.com/m3db/m3/src/x/io"
 	"github.com/m3db/m3/src/x/pool"
 )
 
 type decoder struct {
 	r              io.Reader
+	rOpts          xio.ResettableReaderOptions
+	rr             xio.ResettableReader
 	buffer         []byte
 	bytesPool      pool.BytesPool
 	maxMessageSize int
+	opts           Options
 }
 
 // NewDecoder decodes a new decoder, the implementation is not thread safe.
-func NewDecoder(r io.Reader, opts Options) Decoder {
+func NewDecoder(r io.Reader, opts Options, bufferSize int) Decoder {
 	if opts == nil {
 		opts = NewOptions()
 	}
 	pool := opts.BytesPool()
+	rOpts := xio.ResettableReaderOptions{ReadBufferSize: bufferSize}
 	return &decoder{
 		r:              r,
```

Comment:
Maybe we can even get rid of this base reader on the struct since we functionally are wrapping it now?